### ISSUE TITLE
Prestashop ignores accept-language request header

### ISSUE THREAD

**Issue by ProteanCode** [PrestaShop/PrestaShop#17258]
**Describe the bug**
Prestashop ignores the accept-language header even if the option to detect language from browser has been choosen.

I have 2 language packs there, PL and GB

The shop url is https://arcanastones.com/ and it always load in GB language 

**To Reproduce**
1. Install any working extension that allows you to manipulate accept-language, this one works fine:
https://chrome.google.com/webstore/detail/advanced-page-language-sw/mhgjlbolkmcookadjpolimbddngjgbol

You can also specify the language in advanced Chrome settings

2. Visit the page with language set to Polish (pl)

You will still see the GB language selected instead of PL

**Additional information**
PrestaShop version: 1.7.6.1
PHP version: 7.2

**Comment by khouloudbelguith** [comment]
Hi @ProteanCode,
I did not manage to reproduce the issue with PS1.7.6.3.
I attached a screen record
https://drive.google.com/file/d/1QJdmSx-IcQOPNrGKVy1z3kF8d6XrWHnV/view
In your case, when GB is selected => the language is not added to the URL like this for example: `http://prestashop.test/ps/1763/gb/`

It is Ok with PL language.
Did you make any changes in the URLs or did you use a custom module?

Thanks!

**Comment by khouloudbelguith** [comment]
Simlar issue: https://github.com/PrestaShop/PrestaShop/issues/10744

**Comment by ProteanCode** [comment]
Why Tools::setCookieLanguage is so complex?

According to MDN

> The Accept-Language request HTTP header advertises which languages the client is able to understand, and which locale variant is preferred.

I cannot understand this code at all
https://github.com/PrestaShop/PrestaShop/blob/6195e42cfeb10036f1845d3319ec2159b5284e4a/classes/Tools.php#L572

This should be as simple as "if language has not been manually set by user, then iterate over languages in HTTP_ACCEPT_LANGUAGE header for finding first possible. If none matched, return default language"

I am digging it further but got dizzied

**Comment by ProteanCode** [comment]
The Tools::getValue('isolang') returns "something" but it should not.

I have no idea how the $_GET array contains data like isolang and id_lang. My web browser did not sent this data

```
*DEBUG*         v1.7.6.1        2020/02/06 - 22:25:03: GET ARRAY
*DEBUG*         v1.7.6.1        2020/02/06 - 22:25:03: Array
(
    [isolang] => ga
    [id_lang] => 5
)

```

**Comment by ProteanCode** [comment]
~~At some point I got a HTTP_ACCEPT_LANGUAGE starting from "en-PL" which does not seems to be a real thing (there is no a Polish variation of English language)
At other point it just starts from "pl".~~

~~Even if I get the correct iso code I am not seeing the page in that language~~

No longer truth, this were data rendered by other user in log files.

**Comment by ProteanCode** [comment]
Sometimes the HTTP_ACCEPT_LANGUAGE is not set.

I would not bother with it since it could be a request to some .map files which probably ignores content negotiation (I didn't log the request URI)

**Comment by ProteanCode** [comment]
getLanguageByIETFCode works wrong.

It fetches languages from ps_lang table which can be modified by user in admin panel (/index.php/improve/international/languages).

My web browser (Brave, a Chrome-based one) is sending my HTTP_ACCEPT_LANGUAGE as "pl", not "pl-PL" and I did not have a "pl" value in "language_code" column of "ps_lang" table.

I updated all the "ga-GA" and "pl-PL" fields into "ga" and "pl" accordingly and it properly handles the HTTP_ACCEPT_LANGUAGE header. The language is now correctly picked, but the language switcher does not work.

I had to modify the code inside Tools.php class by adding that else statement (which is basically "i dont care, just use that header if your logic fails")

```
if (!Tools::getValue('isolang') && !Tools::getValue('id_lang') && (!$cookie->id_lang || isset($cookie->detect_language))
            && isset($_SERVER['HTTP_ACCEPT_LANGUAGE'])) {
            $array = explode(',', Tools::strtolower($_SERVER['HTTP_ACCEPT_LANGUAGE']));
            $string = $array[0];

            if (Validate::isLanguageCode($string)) {
                $lang = Language::getLanguageByIETFCode($string);
                if (Validate::isLoadedObject($lang) && $lang->active && $lang->isAssociatedToShop()) {
                    Context::getContext()->language = $lang;
                    $cookie->id_lang = (int) $lang->id;
                }
            }
        } else {
            if(isset($_SERVER['HTTP_ACCEPT_LANGUAGE'])) {
                $array = explode(',', Tools::strtolower($_SERVER['HTTP_ACCEPT_LANGUAGE']));
                $string = $array[0];

                if (Validate::isLanguageCode($string)) {
                    $lang = Language::getLanguageByIETFCode($string);
                    if (Validate::isLoadedObject($lang) && $lang->active && $lang->isAssociatedToShop()) {
                        Context::getContext()->language = $lang;
                        $cookie->id_lang = (int) $lang->id;
                    }
                }
            }
        }
```

Now the language switcher does not work, but I am too tired to dig it further now.

My browsers sends "pl", I see in logs that other browser sent "en-pl".

Header-based language detection is messed up because of the browsers implementation or something.

**Comment by khouloudbelguith** [comment]
@ProteanCode,
Did you make any changes in the URLs or did you use a custom module?
We need to retrieve the PHP error log and the debug mode report in order to find out what's wrong.
Don't you know how to get this information? Please read the following article:
http://build.prestashop.com/howtos/misc/how-to-create-bug-report/

Thanks!

**Comment by ProteanCode** [comment]
> @ProteanCode,
> Did you make any changes in the URLs or did you use a custom module?
> We need to retrieve the PHP error log and the debug mode report in order to find out what's wrong.
> Don't you know how to get this information? Please read the following article:
> http://build.prestashop.com/howtos/misc/how-to-create-bug-report/
> 
> Thanks!

I will try to reproduce the bug on fresh prestashop instance, but I will take on this later (like 9PM CET)

**Comment by khouloudbelguith** [comment]
@ProteanCode, thanks!
Waiting for your feedback.

**Comment by ProteanCode** [comment]
Seems to work just fine on localhost with basic template on dockerized instance even proxied through nginx. I will need to check for more

**Comment by ProteanCode** [comment]
I really don't like the design of Prestashop codebase, I had to read through debug_backtrace just to see that `override` word in a path.

So from what I see the plugins may override the core prestashop classes, one of these classes is `Dispatcher.php` and it has indeed been overriden. This explains why I could not rely on `setRequestUri` method in core Dispatcher.

I can see that setRequestUri is overriden by x13links module

**Comment by ProteanCode** [comment]
The **x13links** module has been removing the isocode from URL, I had to disable its variable (`SEOURL_REMOVE_DEFAULT_ISO`) by switching a field in admin panel.

I don't really consider it a bug in prestashop core, but rather misleading module settings, **x13links** authors should really make a note that removing isocode from url breaks language autodetection.

From what I saw their addon is not in prestashop marketplace, they have their own page https://x13.pl/moduly-prestashop/przyjazne-linki-bez-id.html where they sell their modules encrypted / compiled with ionCube.

@khouloudbelguith In some way you gave me that tip in very first comments, but I was not able to understand how important it is for language autodetecting. Now everything makes sense
